File: products/tutorputor/services/tutorputor-platform/src/clients/feature-store.client.ts
/**
 * Feature Store Platform Client
 *
 * Optional singleton for ingesting and retrieving feature vectors from the
 * Ghatana platform Feature Store. Fully optional — when FEATURE_STORE_URL is
 * absent every call is a no-op (ingestAsync logs and returns, getFeatures
 * returns an empty object).
 *
 * Usage:
 *   import { featureStoreClient } from './feature-store.client.js';
 *   // Fire-and-forget — does not block the request path
 *   featureStoreClient.ingestAsync('acme', 'student-42', { session_count: 5 });
 *   // Awaitable
 *   await featureStoreClient.ingest('acme', 'student-42', { score: 0.87 });
 *   // Read features
 *   const features = await featureStoreClient.getFeatures('acme', 'student-42', ['score']);
 *
 * Environment variables:
 *   FEATURE_STORE_URL — base URL, e.g. http://feature-store:8080
 */

const REQUEST_TIMEOUT_MS = 5_000;

class FeatureStoreClient {
  private static instance: FeatureStoreClient;
  private readonly baseUrl: string | undefined;

  private constructor() {
    this.baseUrl = process.env['FEATURE_STORE_URL'];
    if (!this.baseUrl) {
      console.warn('[FeatureStoreClient] FEATURE_STORE_URL not configured — feature store is disabled');
    }
  }

  static getInstance(): FeatureStoreClient {
    if (!FeatureStoreClient.instance) {
      FeatureStoreClient.instance = new FeatureStoreClient();
    }
    return FeatureStoreClient.instance;
  }

  /**
   * Ingest features asynchronously (fire-and-forget).
   * Errors are swallowed and logged as warnings so the caller is never blocked.
   */
  ingestAsync(tenantId: string, entityId: string, features: Record<string, unknown>): void {
    this.ingest(tenantId, entityId, features).catch((err: unknown) => {
      console.warn(
        `[FeatureStoreClient] ingestAsync failed for entity ${entityId}: ` +
        `${err instanceof Error ? err.message : String(err)}`,
      );
    });
  }

  /**
   * Ingest features synchronously (awaitable).
   * Resolves without error when the feature store is unconfigured.
   */
  async ingest(
    tenantId: string,
    entityId: string,
    features: Record<string, unknown>,
  ): Promise<void> {
    if (!this.baseUrl || Object.keys(features).length === 0) return;

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
    try {
      const res = await fetch(
        `${this.baseUrl}/api/v1/features/${encodeURIComponent(tenantId)}/${encodeURIComponent(entityId)}/batch`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ features }),
          signal: controller.signal,
        },
      );
      if (!res.ok) {
        throw new Error(`HTTP ${res.status}`);
      }
    } finally {
      clearTimeout(timeoutId);
    }
  }

  /**
   * Retrieve named features for an entity.
   * Returns an empty object when the feature store is unconfigured or unavailable.
   */
  async getFeatures(
    tenantId: string,
    entityId: string,
    featureNames: string[],
  ): Promise<Record<string, unknown>> {
    if (!this.baseUrl || featureNames.length === 0) return {};

    const params = new URLSearchParams();
    featureNames.forEach((n) => params.append('names', n));
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
    try {
      const res = await fetch(
        `${this.baseUrl}/api/v1/features/${encodeURIComponent(tenantId)}/${encodeURIComponent(entityId)}?${params}`,
        { signal: controller.signal },
      );
      if (!res.ok) {
        console.warn(`[FeatureStoreClient] getFeatures returned HTTP ${res.status}`);
        return {};
      }
      return (await res.json()) as Record<string, unknown>;
    } catch (err: unknown) {
      console.warn(
        `[FeatureStoreClient] getFeatures failed: ${err instanceof Error ? err.message : String(err)}`,
      );
      return {};
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

export const featureStoreClient = FeatureStoreClient.getInstance();

File: products/audio-video/libs/common/src/main/java/com/ghatana/audio/video/common/platform/FeatureStoreClient.java
package com.ghatana.audio.video.common.platform;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @doc.type class
 * @doc.purpose Optional HTTP client for the Ghatana Feature Store.
 *              Allows audio-video services to asynchronously record inference
 *              events (transcriptions, detections, voice generations) as feature
 *              data for downstream ML training and analytics.
 * @doc.layer product
 * @doc.pattern Service
 *
 * <p>Configuration: set {@code FEATURE_STORE_URL} to the platform Feature Store
 * base URL (e.g. {@code http://feature-store:8080}). When absent every ingest
 * call is a silent no-op — the inference path is never blocked.
 *
 * <p>All ingest calls are fire-and-forget: a background thread sends the HTTP
 * request; exceptions are swallowed and logged as WARN.
 */
public final class FeatureStoreClient {

    private static final Logger LOG = Logger.getLogger(FeatureStoreClient.class.getName());
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);

    private static volatile FeatureStoreClient instance;

    private final String baseUrl;
    private final HttpClient http;
    private final java.util.concurrent.ExecutorService bgExecutor;

    private FeatureStoreClient(String baseUrl) {
        this.baseUrl = baseUrl;
        this.http = HttpClient.newBuilder()
                .connectTimeout(REQUEST_TIMEOUT)
                .build();
        this.bgExecutor = java.util.concurrent.Executors.newSingleThreadExecutor(r -> {
            Thread t = new Thread(r, "av-feature-store-ingest");
            t.setDaemon(true);
            return t;
        });
    }

    /**
     * Return the singleton instance, configured from {@code FEATURE_STORE_URL}.
     */
    public static FeatureStoreClient getInstance() {
        if (instance == null) {
            synchronized (FeatureStoreClient.class) {
                if (instance == null) {
                    String url = System.getenv("FEATURE_STORE_URL");
                    if (url == null || url.isBlank()) {
                        LOG.warning("[FeatureStoreClient] FEATURE_STORE_URL not set — feature ingest disabled");
                        url = "";
                    }
                    instance = new FeatureStoreClient(url);
                }
            }
        }
        return instance;
    }

    /**
     * Ingest a set of features for an entity (fire-and-forget).
     * Returns immediately; the HTTP call is made on a background daemon thread.
     *
     * @param tenantId   Tenant identifier.
     * @param entityId   Entity identifier (e.g. user ID, session ID).
     * @param features   Key-value pairs to record.
     */
    public void ingestAsync(String tenantId, String entityId, Map<String, Object> features) {
        if (baseUrl.isBlank() || features.isEmpty()) return;
        bgExecutor.execute(() -> {
            try {
                ingest(tenantId, entityId, features);
            } catch (Exception e) {
                LOG.log(Level.WARNING, "[FeatureStoreClient] ingestAsync failed for entity " + entityId, e);
            }
        });
    }

    /**
     * Ingest features synchronously. Callers should prefer {@link #ingestAsync} on
     * the hot inference path.
     *
     * @param tenantId  Tenant identifier.
     * @param entityId  Entity identifier.
     * @param features  Feature map to persist.
     * @throws IOException          On network or server error.
     * @throws InterruptedException If the calling thread is interrupted.
     */
    public void ingest(String tenantId, String entityId, Map<String, Object> features)
            throws IOException, InterruptedException {
        if (baseUrl.isBlank() || features.isEmpty()) return;

        String body = toJson(features);
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + "/api/v1/features/"
                        + encode(tenantId) + "/" + encode(entityId) + "/batch"))
                .timeout(REQUEST_TIMEOUT)
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                .header("Content-Type", "application/json")
                .build();

        HttpResponse<Void> resp = http.send(req, HttpResponse.BodyHandlers.discarding());
        if (resp.statusCode() >= 400) {
            throw new IOException("Feature Store returned HTTP " + resp.statusCode());
        }
    }

    /** Minimal JSON object serializer for primitive values only. */
    private static String toJson(Map<String, Object> map) {
        StringBuilder sb = new StringBuilder("{\"features\":{");
        boolean first = true;
        for (Map.Entry<String, Object> e : map.entrySet()) {
            if (!first) sb.append(',');
            sb.append('"').append(e.getKey().replace("\"", "\\\"")).append("\":");
            Object v = e.getValue();
            if (v instanceof Number || v instanceof Boolean) {
                sb.append(v);
            } else {
                sb.append('"').append(String.valueOf(v).replace("\"", "\\\"")).append('"');
            }
            first = false;
        }
        sb.append("}}");
        return sb.toString();
    }

    private static String encode(String s) {
        return java.net.URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}
